### ISSUE TITLE
Rethink `tar_renv()`

### ISSUE THREAD

**Issue by wlandau** [wlandau/targets#224]
## Prework

* [x] Read and agree to the [code of conduct](https://github.com/wlandau/targets/blob/main/CODE_OF_CONDUCT.md) and [contributing guidelines](https://github.com/wlandau/targets/blob/main/CONTRIBUTING.md).
* [x] If there is [already a relevant issue](https://github.com/wlandau/targets/issues), whether open or closed, comment on the existing thread instead of posting a new issue.
* [ ] For any problems you identify, post a [minimal reproducible example](https://www.tidyverse.org/help/) so the maintainer can troubleshoot. A reproducible example is:
    * [ ] **Runnable**: post enough R code and data so any onlooker can create the error on their own computer.
    * [ ] **Minimal**: reduce runtime wherever possible and remove complicated details that are irrelevant to the issue at hand.
    * [ ] **Readable**: format your code according to the [tidyverse style guide](https://style.tidyverse.org/).

## Problems

 I have tried to use `tar_renv()` in a few different projects now, but it still has not stuck for me. It has a couple problems:

1. Specialized formats (e.g. `tar_target(format = "qs")`) and functions like `tar_make_clustermq()` and `tar_visnetwork()` can incur additional package dependencies.
2. It is awkward to call `tar_env()` inside `_targets.R` because it requires that we run `tar_make()` or a utility like `tar_manifest()` before we can create the `renv` library.

## Proposal

1. `tar_renv()` could spin up a new `callr` process and source `_targets.R` just like helpers `tar_visnetwork()` and `tar_manifest()`.
2. Always account for `tar_option_get("packages")` without requiring it as input to the `packages` argument.
3. Automatically include the format-specific packages required for the workflow. That means looking at `tar_option_get("format")` and crawling through `$settings$format` for each target object in the pipeline.
4. Support a single `extras` argument for extra packages. Should default to `c("clustermq", "future", "rstudioapi", "visNetwork")`. So the overall signature of `tar_renv()` should be really simple: `tar_renv <- function(extras = c("clustermq", "future", "rstudioapi", "visNetwork")) {...}` 
5. Users are likely to have `tar_renv()` inside `_targets.R`, which will cause infinite recursion after (1). `assert_target_script()` should add to the existing guardrails:

https://github.com/wlandau/targets/blob/87ba2369c186dbd00d7dd5d33839eb05590c3e72/R/utils_assert.R#L227

@mattwarkentin, what do you think? Since you wrote `tar_renv()`, are you interested in taking on the implementation?

**Comment by mattwarkentin** [comment]
@wlandau I'm definitely interested in taking this on. Let me think over these comments and then I'll try to get started on a PR draft. Happy to rethink this function, I want the API to be friction-free and intuitive, and it sounds like it isn't in its current form.

**Comment by wlandau** [comment]
Thanks, Matt! The API itself was fine, but the usage was a little awkward because of the packages it left out. 

If it helps, [`tar_prune()`](https://github.com/wlandau/targets/blob/main/R/tar_prune.R) is a good way to model the new `tar_env()`. The user-side function should prepare the arguments and submit a background process:

```r
tar_renv <- function(
  extras = c("clustermq", "future", "rstudioapi", "visNetwork"),
  callr_function = callr::r,
  callr_arguments = list()
) {
  assert_target_script()
  assert_chr(extras, "extras must be a character vector")
  assert_callr_function(callr_function)
  assert_list(callr_arguments, "callr_arguments mut be a list.")
  callr_outer(
    targets_function = tar_renv_inner,
    targets_arguments = list(extras = extras),
    callr_function = callr_function,
    callr_arguments = callr_arguments
  )
  invisible()
}
```

Then, `tar_renv_inner()` is what runs inside the background process.

```r
tar_renv_inner <- function(pipeline, extras) {
  pipeline_validate_lite(pipeline)
  ...
}
```

**Comment by mattwarkentin** [comment]
Just so I'm clear on how the function should be used and what it should do:

- Do you still envision this staying as a function simply included in `_targets.R` or do you envision it as a function called, perhaps interactively, like many of the other side-effect `tar_*` functions outside of `_targets.R` and introspecting all packages used in the pipeline?

- Do you think think the function should create `_packages.R` and return nothing invisibly?

**Comment by wlandau** [comment]
> Do you still envision this staying as a function simply included in _targets.R or do you envision it as a function called interactively outside of _targets.R, like many of the other side-effect tar_* functions, and introspecting all packages used in the pipeline?

Definitely the latter, outside `_targets.R` just like `tar_visnetwork()`.

> Do you think think the function should create _packages.R and return nothing invisibly?

Yes. And I totally forgot about the `path` argument, which I think we should keep.

**Comment by mattwarkentin** [comment]
Okay, agreed. I think I have a good mental model of what we want. I'll start into it.

**Comment by mattwarkentin** [comment]
Okay, I have a decent first draft in place for the proposed functionality in https://github.com/wlandau/targets/issues/224#issue-744092071. 

However, any dependencies explicitly loaded in `_targets.R` (or any other R script for that matter) via `library()`, `require()`, or `pkg::fun` namespaced calls aren't currently captured, unless these packages are also capture through other standard package declarations.

I presume we want to capture these as well, right?

**Comment by wlandau** [comment]
I don’t think we need to. renv should notice those packages using the same process that crawls through _packages.R.

**Comment by mattwarkentin** [comment]
Right, yes. Lost sight of that for a second.

**Comment by wlandau** [comment]
I am working on a little restructuring to make it easier to get the packages given a target's format. Hopefully that will simplify things on your end.

Also, I forgot about individual packages the user supplies to the `packages` argument of `tar_target()`. You can get these with `x$command$packages` for a target object `x`.

**Comment by mattwarkentin** [comment]
I should submit my quick-and-dirty draft PR so we can coordinate. I wrote a hacky way to get the packages for a targets format too.

Regarding the second point, I actually just ignored packages from `tar_option_set()` and only interrogated packages contained in pipeline$targets$x because packages from `tar_option_set()` are propagated to the individual targets.

**Comment by wlandau** [comment]
> I should submit my quick-and-dirty draft PR so we can coordinate. I wrote a hacky way to get the packages for a targets format too.

Sure, that would help.

In recent commits, I added some methods to keep track package requirements more cleanly in case we need to do it again for other reasons. On your end, you should now only need to call `pipeline_get_packages()` on the pipeline object. Sorry if that invalidates some of your efforts.

**Comment by wlandau** [comment]
> Regarding the second point, I actually just ignored packages from tar_option_set() and only interrogated packages contained in pipeline$targets$x because packages from tar_option_set() are propagated to the individual targets.

I guess you're right. `tar_option_set(packages = ...)` really does only serve the individual targets. In fact, if you set `packages` in every single `tar_target()` call, `tar_option_get("packages")` will be completely ignored. But I wonder, should we include it anyway? What if the user decides to add more targets without supplying explicit package requirements and defaulting to `tar_option_get("packages")` incurs new pkg deps?

**Comment by wlandau** [comment]
I guess by that logic we also need the package corresponding to `tar_option_get("format")`. You can get it with `store_get_packages(store_init(tar_option_get("format")))`

**Comment by mattwarkentin** [comment]
> On your end, you should now only need to call pipeline_get_packages() on the pipeline object. Sorry if that invalidates some of your efforts.

Haha, no worries. I actually had already created a function with the exact same name (and presumably functionality). At least we are thinking the same way.

> What if the user decides to add more targets without supplying explicit package requirements and defaulting to `tar_option_get("packages")` incurs new pkg deps?

If a user adds new targets without adding target-specific package reqs, won't this target just get whatever packages are propagated from `tar_option_set()`?

Let me submit this PR draft, so we can review our current setup and think through what needs to be done.

**Comment by wlandau** [comment]
> If a user adds new targets without adding target-specific package reqs, won't this target just get whatever packages are propagated from tar_option_set()?

Yes, but that could be the first time `tar_option_get("packages")` actually comes into play, and it would require another call to `tar_renv()` after the targets are added. If we account for `tar_option_get("packages")`, the first call to `tar_renv()` already prepares for those new targets, so `tar_renv()` doesn't need to be called again.

**Comment by wlandau** [comment]
Fixed in #229.
